website/facedetection/models.py

```diff
@@ -0,0 +1,332 @@
+import os
+from secrets import token_urlsafe
+
+from django.db import models
+from django.db.models import Count, IntegerField, Value
+from django.db.models.functions import Coalesce
+
+from queryable_properties.managers import QueryablePropertiesManager
+from queryable_properties.properties import AnnotationProperty
+from thumbnails.fields import ImageField
+
+from members.models import Member
+from photos.models import Photo
+
+
+class FaceDetectionUser(Member):
+    class Meta:
+        proxy = True
+
+
+def authentication_token() -> str:
+    """Generate a 40 characters long base64 token suitable for authentication."""
+    return token_urlsafe(30)
+
+
+def reference_face_uploadto(instance, filename):
+    """Get path of file to upload to."""
+    random = token_urlsafe(6)
+    extension = os.path.splitext(filename)[1]
+    return os.path.join(
+        "facedetection/reference-faces",
+        f"{instance.user.username}-{random}{extension}",
+    )
+
+
+class BaseFaceEncodingSource(models.Model):
+    """Abstract model for a source of face encodings."""
+
+    class Status(models.TextChoices):
+        PROCESSING = "processing"
+        DONE = "done"
+        REJECTED = "rejected"
+
+    status = models.CharField(
+        max_length=16,
+        choices=Status.choices,
+        default=Status.PROCESSING,
+        help_text="Status of the encoding extraction process.",
+    )
+
+    token = models.CharField(
+        max_length=40,
+        default=token_urlsafe,
+        editable=False,
+        help_text="Token used by a Lambda to authenticate "
+        "to the API to submit encoding(s) for this source.",
+    )
+
+    submitted_at = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The time at which the source was "
+        "submitted to the Lambda for processing.",
+    )
+
+    class Meta:
+        abstract = True
+
+
+class FaceDetectionPhoto(BaseFaceEncodingSource):
+    """A source of face encodings from a Photo."""
+
+    photo = models.OneToOneField(Photo, on_delete=models.CASCADE)
+
+    num_faces = AnnotationProperty(
+        Coalesce(Count("encodings"), Value(0), output__field=IntegerField())
+    )
+
+    objects = QueryablePropertiesManager()
+
+    def __str__(self):
+        return f"{self.photo.album} - {self.photo}"
+
+
+class ReferenceFace(BaseFaceEncodingSource):
+    """A source of face encodings from a reference photo of a user's face.
+
+    If a user marks a reference face for deletion, it will be kept for some
+    time to allow us to monitor whether people searched for faces of others.
+    """
+
+    user = models.ForeignKey(
+        FaceDetectionUser,
+        on_delete=models.CASCADE,
+        related_name="reference_faces",
+    )
+
+    file = ImageField(upload_to=reference_face_uploadto)
+
+    created_at = models.DateTimeField(auto_now_add=True, editable=False)
+    marked_for_deletion_at = models.DateTimeField(null=True, blank=True)
```

Comment:
Why is this a separate field? We always compute this right?

Reply:
marked_for_deletion_at? What would we be able to compute that from?

Reply:
Nvm, I forgot for what we use them.